File: backoffice/FondaLogic/BaseCommand.cs
using System;
using com.ds201625.fonda.DataAccess.FactoryDAO;

namespace com.ds201625.fonda.Logic
{
	public abstract class BaseCommand
	{
		
		private Parameter[] _parameters;

		private Object _result;

		public BaseCommand ()
		{
			_parameters = InitParameters ();
			_result = null;
		}

		public Object Result
		{
			get { return _result; }
			protected set { _result = value; }
		}

		public void SetParameter(int index, Object data)
		{
			if ( index < 0 || index >= _parameters.Length)
				// TODO: Exeption personalizada
				throw new Exception("Index (" + index + " fuera de rango." );

			_parameters [index].Data = data;
		}

		public void  Run ()
		{
			ValidateParameters ();
			Invoke ();
		}

		protected Object GetParameter(int index)
		{
			if ( index < 0 || index >= _parameters.Length)
				// TODO: Exeption personalizada
				throw new Exception("Index (" + index + " fuera de rango." );
			return _parameters [index].Data;
		}

		protected abstract Parameter[] InitParameters ();

		protected abstract void Invoke();

		private void ValidateParameters()
		{
			for (int i = 0; i < _parameters.Length; i++)
			{
				if (_parameters[i].IsRequiered() && _parameters[i].Data == null)
					// TODO: Exeption personalizada
					throw new Exception("Parametro requerido no espesificado index (" + i + ")");
			}
		}

		protected FactoryDAO FacDao
		{
			get { return FactoryDAO.Intance; }
		}
	}
}

File: frontoffice/app/src/main/java/com/ds201625/fonda/logic/BaseCommand.java
package com.ds201625.fonda.logic;

import com.ds201625.fonda.data_access.retrofit_client.exceptions.AddFavoriteRestaurantFondaWebApiControllerException;
import com.ds201625.fonda.data_access.retrofit_client.exceptions.DeleteFavoriteRestaurantFondaWebApiControllerException;
import com.ds201625.fonda.data_access.retrofit_client.exceptions.FindFavoriteRestaurantFondaWebApiControllerException;
import com.ds201625.fonda.data_access.retrofit_client.exceptions.GetAllRestaurantsFondaWebApiControllerException;

/**
 * Base para los comandos
 */
public abstract class BaseCommand implements Command {

    /**
     * Parametos
     */
    private Parameter[] parameters;

    /**
     * Para contener el resultado del la ejecucion
     */
    private Object result;

    /**
     * Constructor unico
     */
    public BaseCommand() {
        this.result = null;
        this.parameters = this.setParameters();
    }

    /**
     * Obtiene el resultado
     * @return resultado
     */
    public Object getResult() {
        return  this.result;
    }

    /**
     * Asigna el resultado
     * @param data
     */
    protected void setResult(Object data) {
        this.result = data;
    }

    /**
     * Invocacion a la ejecuncion del comando
     * @throws Exception
     */
    @Override
    public void run() throws Exception {
        this.validateParameters();
        this.invoke();
    }

    /**
     * Asignar parametros
     * @param index posición del parametro
     * @param data objeto a contener
     * @throws Exception
     */
    @Override
    public void setParameter(int index, Object data)
            throws ParameterOutOfIndexException,InvalidParameterTypeException {

        if (this.parameters == null) {
            return;
        }

        if ( index < 0 || index >= this.parameters.length)
            throw ParameterOutOfIndexException.generate(this.getClass().toString(),index);

        parameters[index].setData(data);
    }

    /**
     * Obtener un parametro
     * @param index posicion
     * @return Objeto que contiene el parametro
     * @throws Exception
     */
    protected Object getParameter(int index) throws ParameterOutOfIndexException {
        if (this.parameters == null)
            throw ParameterOutOfIndexException.generate(this.getClass().toString(),index);

        if ( index < 0 || index >= this.parameters.length)
            throw ParameterOutOfIndexException.generate(this.getClass().toString(),index);

        return parameters[index].getData();
    }

    /**
     * Para la inicializacion de los parametros
     * @return
     */
    protected abstract Parameter[] setParameters();

    /**
     * Invocacion de la acción del comando
     */
    protected abstract void invoke() throws Exception;


    /**
     * Valida los parametros requeridos
     * @throws Exception
     */
    private void validateParameters() throws EmptyRequieredParameterException {
        if (this.parameters == null) {
            return;
        }
        for (int i = 0; i < parameters.length; i++) {
            if (parameters[i].isRequiered() && parameters[i].getData() == null)
                throw EmptyRequieredParameterException.generate(this.getClass().toString(),i);
        }
    }
}

